Source/WebCore/rendering/svg/RenderSVGImage.cpp

```diff
@@ -0,0 +1,413 @@
+        imageResource().cachedImage()->addClientWaitingForAsyncDecoding(*this);
+
+    return drawResult;
+}
+
+void RenderSVGImage::paintForeground(PaintInfo& paintInfo, const LayoutPoint& paintOffset)
+{
+    GraphicsContext& context = paintInfo.context();
+    if (context.invalidatingImagesWithAsyncDecodes()) {
+        if (cachedImage() && cachedImage()->isClientWaitingForAsyncDecoding(*this))
+            cachedImage()->removeAllClientsWaitingForAsyncDecoding();
+        return;
+    }
+
+    if (!imageResource().cachedImage()) {
+        page().addRelevantUnpaintedObject(this, visualOverflowRectEquivalent());
+        return;
+    }
+
+    RefPtr<Image> image = imageResource().image();
+    if (!image || image->isNull()) {
+        page().addRelevantUnpaintedObject(this, visualOverflowRectEquivalent());
+        return;
+    }
+
+    FloatRect contentBoxRect = borderBoxRectEquivalent();
+    FloatRect replacedContentRect(0, 0, image->width(), image->height());
+    imageElement().preserveAspectRatio().transformRect(contentBoxRect, replacedContentRect);
+
+    contentBoxRect.moveBy(paintOffset);
+
+    ImageDrawResult result = paintIntoRect(paintInfo, contentBoxRect, replacedContentRect);
+
+    if (cachedImage()) {
+        // For now, count images as unpainted if they are still progressively loading. We may want
+        // to refine this in the future to account for the portion of the image that has painted.
+        FloatRect visibleRect = intersection(replacedContentRect, contentBoxRect);
+        if (cachedImage()->isLoading() || result == ImageDrawResult::DidRequestDecoding)
+            page().addRelevantUnpaintedObject(this, enclosingLayoutRect(visibleRect));
+        else
+            page().addRelevantRepaintedObject(this, enclosingLayoutRect(visibleRect));
+    }
+}
+
+bool RenderSVGImage::nodeAtPoint(const HitTestRequest& request, HitTestResult& result, const HitTestLocation& locationInContainer, const LayoutPoint& accumulatedOffset, HitTestAction hitTestAction)
+{
+    if (hitTestAction != HitTestForeground)
+        return false;
+
+    auto adjustedLocation = accumulatedOffset + currentSVGLayoutLocation();
+
+    auto visualOverflowRect = visualOverflowRectEquivalent();
+    visualOverflowRect.moveBy(adjustedLocation);
+    if (!locationInContainer.intersects(visualOverflowRect))
+        return false;
+
+    auto localPoint = locationInContainer.point();
+    auto boundingBoxTopLeftCorner = flooredLayoutPoint(objectBoundingBox().minXMinYCorner());
+    auto coordinateSystemOriginTranslation = boundingBoxTopLeftCorner - adjustedLocation;
+    localPoint.move(coordinateSystemOriginTranslation);
+
+    if (!SVGRenderSupport::pointInClippingArea(*this, localPoint))
+        return false;
+
+    PointerEventsHitRules hitRules(PointerEventsHitRules::SVG_IMAGE_HITTESTING, request, style().pointerEvents());
+    bool isVisible = (style().visibility() == Visibility::Visible);
```

Comment:
Bit of a preference thing, but not sure whether outer parentheses or the whole variable isVisible are needed.

Reply:
Not strictly, depends on which operator has precedence. I often go for explicit braces....